src/lib/merkle_mask/masking_merkle_tree.ml

```diff
@@ -944,12 +1034,24 @@ module Make (Inputs : Inputs_intf.S) = struct
     let location_of_sexp = Location.t_of_sexp
   end
 
-  let set_parent t parent =
+  let set_parent ?accumulated:accumulated_opt t parent =
     assert (Result.is_error t.parent) ;
     assert (Option.is_none (Async.Ivar.peek t.detached_parent_signal)) ;
     assert (Int.equal t.depth (Base.depth parent)) ;
     t.parent <- Ok parent ;
     t.current_location <- Attached.last_filled t ;
+    (* If [t.accumulated] isn't empty, then this mask had a parent before
+       and now we just reparent it (which may only happen if both old and new parents
+        have the same merkle root (and some masks in between may have been removed),
+       hence no need to modify [t.accumulated]) *)
+    ( match accumulated_opt with
+    | Some { current; next; base; detached_next_signal }
+      when Option.is_none t.accumulated ->
+        maps_merge current t.maps ;
```

Comment:
This mutably merges the contents of `t.maps` into the values passed in from `accumulated_opt`. I don't think that's what we want.

Reply:
I'm addressing this one in my `update_maps` refactor. I left the other issue I pointed out [here](https://github.com/MinaProtocol/mina/pull/14617#discussion_r1418092944) till later, as it requires more changes to implement.